lib/stove-testing-e2e-elasticsearch/src/main/kotlin/com/trendyol/stove/testing/e2e/elasticsearch/ElasticsearchSystem.kt

```diff
@@ -0,0 +1,181 @@
+import org.elasticsearch.action.get.GetRequest
+import org.elasticsearch.action.index.IndexRequest
+import org.elasticsearch.client.Request
+import org.elasticsearch.client.RequestOptions
+import org.elasticsearch.client.RestClient
+import org.elasticsearch.client.RestClientBuilder
+import org.elasticsearch.client.RestHighLevelClient
+import org.elasticsearch.common.xcontent.XContentType
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.testcontainers.elasticsearch.ElasticsearchContainer
+import kotlin.reflect.KClass
+
+data class ElasticsearchSystemOptions(
+    val configureExposedConfiguration: (ElasticSearchExposedConfiguration) -> List<String> = { _ -> listOf() },
+    val jsonSerializer: StoveJsonSerializer = StoveJacksonJsonSerializer(),
+)
+
+fun TestSystem.withElasticsearch(
+    index: String,
+    registry: String = "docker.elastic.co/",
+    imageVersion: String = "7.10.2",
+    options: ElasticsearchSystemOptions = ElasticsearchSystemOptions(),
+): TestSystem {
+    val elasticsearchContainer =
+        withProvidedRegistry("elasticsearch/elasticsearch:$imageVersion", registry) {
+            ElasticsearchContainer(it)
+        }
+    elasticsearchContainer.addExposedPorts(9200)
+    this.getOrRegister(
+        ElasticsearchSystem(this, ElasticsearchContext(index, elasticsearchContainer, options))
+    )
+    return this
+}
+
+data class ElasticSearchExposedConfiguration(
+    val host: String,
+    val port: Int,
+)
+
+data class ElasticsearchContext(
+    val index: String,
+    val container: ElasticsearchContainer,
+    val options: ElasticsearchSystemOptions,
+)
+
+fun TestSystem.elasticsearch(): ElasticsearchSystem =
+    getOrNone<ElasticsearchSystem>().getOrElse {
+        throw SystemNotRegisteredException(ElasticsearchSystem::class)
+    }
+
+class ElasticsearchSystem internal constructor(
+    override val testSystem: TestSystem,
+    private val context: ElasticsearchContext,
+) : DocumentDatabaseSystem, RunAware, ExposesConfiguration {
+    private lateinit var client: RestHighLevelClient
+
+    private lateinit var exposedConfiguration: ElasticSearchExposedConfiguration
+    private val objectMapper: StoveJsonSerializer = context.options.jsonSerializer
+
+    private val logger: Logger = LoggerFactory.getLogger(javaClass)
+
+    object Constants {
+        const val ONE_MINUTE = 60 * 1000
+        const val ONE_SECOND = 1000
+    }
+
+    override suspend fun run() {
+        context.container.start()
+        exposedConfiguration = ElasticSearchExposedConfiguration(
+            context.container.host,
+            context.container.firstMappedPort
+        )
+        client = createCluster(exposedConfiguration)
+    }
+
+    override suspend fun stop() {
+        context.container.stop()
+    }
+
+    override suspend fun <T : Any> shouldQuery(
+        query: String,
+        assertion: (List<T>) -> Unit,
+        clazz: KClass<T>,
+    ): DatabaseSystem {
```

Comment:
all methods should return `ElasticsearchSystem`

Reply:
Fixed